### DISCUSSION TITLE
essay/2024/04/29/home-server

### DISCUSSION THREAD

**Discussion by giscus** [BenjaminKim/benjaminkim.github.io#207] [Comments]
# essay/2024/04/29/home-server

집에서도 서버를 운영할 수 있을까? 생각은 하지만 실제로 실행하는 사람은 별로 없는 것 같습니다.

https://jeho.page/essay/2024/04/29/home-server.html

**Comment by wapj** [comment]
혹시 아이피는 고정아이피 하셨는지, DDNS 쓰신건지 궁금합니다..ㅎ

**Reply by BenjaminKim** [reply]
DDNS를 쓰지 않았고 고정하지도 않았습니다.
그냥 유동 아이피로 6년 동안 썼습니다.
DHCP라고 해서 아이피가 바뀌질 않더라구요. 모뎀을 껐다 켜야 바뀌는데... 그마저도 안 바뀔 때도 있었습니다.

**Reply by wapj** [reply]
새로운걸 알았네요~ 답변감사합니다~ 👍

**Reply by kgs921** [reply]
개인 유동아이피라고해도 공유기 집안 모뎀으로 인해서는 아이피가 변경되지않습니다. 구성에 따라 다르지만 아파트라면 아파트에들어온 L3스위치 또는 중간 지사의 L4스위치를 재부팅할경우 아이피가 바뀌게됩니다..요즘은 장비들이 좋아져서 아이피 테이블을 새로 구성할만한 작업은 많이없고 이중화등으로 구성되어 단순 재부팅정도만 수행하게되는데 그럴경우 모뎀등에서 아이피 갱신이 이뤄지기 전에 재연결되어 개인사용자는 거의 모른답니다

**Reply by BenjaminKim** [reply]
@kgs921 그렇겠군요. 알려주셔서 감사합니다.

**Comment by voidFunctionReturn0** [comment]
다음 이야기도 매유 기대됩니다.
귀중한 경험 공유해주셔서 감사합니다!

**Reply by BenjaminKim** [reply]
다음 이야기 열심히 적어봐야겠네요. 😁

**Comment by ru-developer** [comment]
혹시 서비스하는데 일반가정집 인터넷 제공 서비스 약관에 위배되지는 않나요? 통신사의 다른 상품으로 가입해야 된다던가 하는 그런 문제가 없는지 궁금합니다.
집에서 서버를 돌려보고 싶은데, 기술 외적으로 어떤 문제들이 있을지 잘 모르겠어서요.

**Reply by BenjaminKim** [reply]
글쎄요. 다른 상품으로 가입하지는 않았습니다.
궁금해서 [약관](https://corp.kt.com/attach/board/BS00000005/70/KT%EC%9D%B8%ED%84%B0%EB%84%B7%EC%84%9C%EB%B9%84%EC%8A%A4%EC%9D%B4%EC%9A%A9%EC%95%BD%EA%B4%80_20200727.pdf)을 좀 찾아보니 아래 내용이 있네요.

② 케이티는 원활한 정보전송 역무를 수행하기 위하여 자율적으로 다음과 같은 경우 해당 역무의
제공을 거부하는 조치를 할 수 있습니다.
1. 웜, 바이러스, 해킹 등으로 인하여 훼손된 컴퓨터에서 다량의 감염의도의 메일전송, 스팸전송
등이 발견된 경우
2. 해당 이용자와 관련하여 국내외로부터 민원 또는 위해 제보가 접수된 경우
3. 기타 케이티와 체결한 약관으로 판단하여 정상적인 정보전송이 아닌 경우로 판단되는
경우(유동IP대역에서 메일서버 등 **인터넷서버 운영**)
4. 기타 고객의 컴퓨터에서 정상적인 정보전송이 아닌 이상트래픽, Looping 등으로 정보전송
역무가 방해된다고 판단되는 경우

서비스 제공을 중지 시킬 수 있다고 적혀 있긴 하나... 아마 그런 일은 쉽게 발생하진 않을 겁니다.
웬만한 서비스의 트래픽은 감지조차 할 수 없을 정도로 미미하기 때문에...(토렌트에 비하면야)

**Reply by iruis** [reply]
만약 가정용 회선이 U+나 SKB인데 광케이블로(흔히 말하는 FTTH) 연결되어있지 않았다면(흔히 말하는 동축 케이블로 집안에 연결 되어있을 경우) 80번 포트같은 서버로 많이 사용되는 포트는 막혀있다고 봐야합니다.

전 그것 때문에 KT씁니다. FTTH가 안들어오는 집이여도 21번, 80번같은 포트가 열려있거든요. KT도 간혹 막힌 지역 있다고 하던데 이러한 상황에 웹서비스를 집에서 돌리고 싶으시다면 80번은 포기해야합니다. 가정용 인터넷으로 신청했으면 약관에 명시되어 있듯 원래는 서버를 운영하면 안되므로 고객센터에 문의해봐도 열어주지 않으니까요.

운영하고 싶으신 서버가 있다면 먼저 포트부터 외부에서 접근 가능한지 확인해보세요. 그리고 인터넷 물린 공유기나 PC가 일주일 이런식으로 전원이 OFF되면 DHCP서비스에서 임대가 만료되어버리는지 아이피가 바뀌고, 거리가 좀 떨어진 곳으로 이사를 하게된다면(약 480미터 거리로 이사했었는데도 아이피가 유지되었습니다) IP가 바뀌니 도메인, DNS(전 dnsever씁니다)서비스 사용해서 도메인을 사용하시길 권하고싶습니다.

**Reply by BenjaminKim** [reply]
경험담 공유 고맙습니다.
저도 KT가 가장 좋더라고요. dnsever 는 예전에 저도 잠깐 쓰다가 지금은 cloudflare 로만 관리하고 있습니다.

**Comment by realrains** [comment]
AWS 비용 아끼려고 서버리스 스택 기웃거리는 것 보다 간단한 서비스에는 젤 확실한 방법같기도 하네요 흥미로운 글 감사합니다!

**Reply by BenjaminKim** [reply]
정말 그렇습니다. 서버리스라고 해서 싸지도 않습니다.

**Comment by Klaus-Schmerz** [comment]
저도 파일 서버가 필요해서 GCP에서 임대해서 쓰다가, 디스크 I/O 비용 보고 얼른 시놀로지로 갈아타버렸죠.
1테라 빌리는데만 한 달에 8만 원~10만 원, 많이 쓰는날엔 하루에도 10만 원이 나왔어요..

전에는 라즈베리파이로 워드프레스도 돌렸는데, 이제는 어떤 서버를 굴려볼지 상상하는게 취미네요 ㅎㅎ
한 쪽 서브컴에는 또 하이퍼바이저 깔아서 가지고 놀고 있거든요

**Reply by trim703** [reply]
https://namu.wiki/w/%ED%99%88%EC%84%9C%EB%B2%84#s-2.3

저는 나무위키 위 문서에 언급된 컴퓨터로 집에서 서버 굴리던 당사자입니다^^
추억이 새록새록하네요~ 지금도 서버는 여전히 굴리고 있고 링크의 서버는 훨씬 업글된 사양으로 IDC에서 저렴하게 돌리고 있습니다(월 33천원).
제게 홈서버의 최대장점은 막대한 용량이었습니다. 지금도 소중한 데이터들이 잘 보존되어 있네요. 최대한 평생에 가깝게 운영할 생각입니다.

**Reply by BenjaminKim** [reply]
두분 다 컴퓨터를 얼마나 좋아하는지 느껴지네요. 너무 좋습니다.

**Reply by Klaus-Schmerz** [reply]
여기는 답글 알림이 따로 없네요
저도 처음 홈서버 시작할 때 기억이 나네요 ㅎㅎ 아파치만 깔면 간단한 웹/파일 서버가 되다니 너무 신기하고..
IDC 입주까지 하시다니 열정이 대단하신 것 같아요.
확실히 용량은 마음대로 정해서 원하는 만큼 쓸 수 있다는게 장점이죠. 아무래도 클라우드는 파일 자체가 통제불능한 서버에 있다는 점도 있고, 이 파일이 감시(?) 당한다 생각할 수도 있으니까요. 또 VPN을 쓴다 한들 외부망에서는 어떤 일이 일어날 지도 모르는 일이구요.
글쓴이님도 답글 달아주시다니 감사하네요 ㅎㅎ

**Comment by kjm1922** [comment]
우연찮게 크롬 Discover 에 떠서 글 잘 읽었습니다.
저도 개인적으로 집에서 24/7 웹서버를 운용중이고 APM 기반으로 가동 하고 있습니다. 지금까지도 별 일이 없고요.

**Reply by BenjaminKim** [reply]
집에서 서버 돌리는 분들이 생각보다 많군요!

**Comment by gaeco0** [comment]
집에서 서버 운영 10년 넘게하다가 사업 시작하면서 사무실로 서버 옮겼는데요 전기세나 인터넷은 별 문제 없습니다.

집에서 운영시에 가장 큰 단점은 저같은 경우 도메인 연결 시 중국 러시아발 해킹 시도구요 보안만 잘 해두면 문제 없습니다

인터넷은 isp에서 가끔 80 443을 차단하는 경우있습니다

**Reply by BenjaminKim** [reply]
해킹 시도는 집이나 aws 나 똑같이 시도하지 않을까요?
아무거나 잡아서 공격하지, 집에서 돌리는 것만 찾아 공격하진 않으니까요.
그래서 그게 가장 큰 단점이라고 할 수는 없겠습니다.

password 로 로그인 하지 못하게 막아 두는게 중요하고 앞 단에 cloudflare 를 두면 좋습니다.

**Reply by gaeco0** [reply]
개인이 운영하면 방화벽도 자기 몫이니까요

**Reply by BenjaminKim** [reply]
보안이 신경이 많이 쓰이긴 합니다. 그래도 cloudflare 가 어려운 일을 많이 해줘서 이제 정말 좋아졌어요.

**Reply by kjm1922** [reply]
저도 Cloudflare 에 DNS 물려서 쓰는 중 입니다.
다만 CF 를 쓰면 TTFB 가 확실히 리전이 타국인 점 때문에
http/2 니 CDN 이니 캐싱이니 별 짓을 다해도 느린건 어쩔 수 없습니다.
CF를 포기하자니 제공해주는 보안 기능이 빠지는걸 감안하면 참고 써야 되나 싶기도 하고요.

**Reply by BenjaminKim** [reply]
맞아요. 한국 리전도 지원해주길 기대하고 있습니다. 조만간 되지 않을까요?

**Comment by Gandy0215** [comment]
저전력 cpu가 장착 된 미니pc를 써도 괜찮겠네요.
20만원 정도 가격에 사이즈도 작으니 부담스럽지도 않고..
최근에 중국에서 해킹프로그램 박은 윈도우 설치해서 판다던데 리눅스 설치할거면 그 걱정도 없겠네요 ㅎ

**Reply by BenjaminKim** [reply]
저전력 미니 pc 는 안 써봐서 모르겠어요. 성능이 너무 안 나오면 만족스럽지 않아서 새로 사게 될수도 있습니다.
저는 20만원 정도짜리 cpu에 램은 가능한 많이 사려고 해요. (그래픽 카드는 안 삽니다)

**Comment by SpookyJelly** [comment]
재호님 글은 항상 재미있습니다. 오랜 기간 쌓인 개발 노하우의 맛을 잘 살려서 글을 쓰시는거 같아요. 만약 K리그 프로그래머가 식당이였다면 못해도 블루리본 10개는 받았을거 같네요 ㅎㅎㅎ

**Reply by BenjaminKim** [reply]
와~ 고맙습니다. 블루리본 10개라니... 최고의 칭찬입니다. 😻

**Comment by Seankim13** [comment]
재호님의 글과 유투브 인터뷰 보면서 많은 영감을 얻고 있습니다. 저는 개발자는 아니지만, 홈서버로 아주 작은 서비스를 준비하고 있는지라 글 하나, 말씀 한마디한마디가 저에게 뼈가 되고 살이 되네요^^ 가시는 길 지켜보면서 항상 응원하겠습니다^^

**Reply by BenjaminKim** [reply]
응원 감사합니다. 힘이나서 더 글을 많이 쓸 수 있을 것 같습니다.

**Comment by Leuconoe** [comment]
저도 최근에 n100미니피씨로 홈 서버 구축했습니다. 레고 만드는 느낌이네요.

**Reply by BenjaminKim** [reply]
이제 홈 서버의 시대가 오는가 봅니다. 😁

**Comment by bluems** [comment]
저도 중학생 때 펜티엄 4로 시작해서, 컴퓨터 바꿀 때마다 서버를 그 부품으로 업글 업글.. 누구도 오지 않고 그땐 그게 더 불편했는데도 뭐가 그리 재밌었는지.. 이 경험 덕분에 일에 치이는게 싫어서 작은 회사 다니지만 리눅스 서버운영 경험만 치면 저도 벌써 16년차네요 ㅋㅋ 우분투 사용으로 치면은 6.04 ? 8.04? 부터 썼는데.. 당시엔 기념품 삼아 우분투 재단에 신청하면 시디 패키지로도 배포되었죠. 영국발이라 악명높은 배송 기간에 잊을 때 쯤 와서 보면 불량률이 30퍼가 넘었던..ㅋㅋㅋㅋ 어머님은 영국에서 시디가 배송오니, 당시 제가 해커라도 된줄 알았다고.. 16년차 홈 서버 운영 중이라니 글 보다가 깨달았습니다.. 지금은 이것저것 내려놓고, 주 용도로 제온 기반 NAS 하나 장만하고 그 안에 도커로 마이크로 서비스 하고 있네요. 지금 이사한 동네는 어째선지 모뎀 켜 있어도 통신사 공사 안내만 떴다하면 스위치도 리셋하는지 아이피 날아가고 그래서(...) ddns도 도커로 살려서 넣고.. 그럼에도 재밌으니 매번 새로운 뭔 또 만들어볼까 하고 살아갑니다~

**Reply by BenjaminKim** [reply]
저도 비슷한 시기부터 [우분투를 사용](https://jeho.page/essay/2022/10/21/ubuntu-life.html) 했습니다.
컴퓨터를 사랑하는 분들이 제 블로그에 많이 찾아주셔서 너무 기쁩니다.

**Comment by haeseong8880** [comment]
저는 개인적으로 가비아 5500원짜리 클라우드 서버 쓰는데 나름 만족하면서 쓰고 있어요 ㅋㅋ

**Reply by BenjaminKim** [reply]
찾아보니 램이 512MB인 것 같은데... 어떤 용도로 쓰고 계시나요?
DB가 필요한 프로그램은 좀 힘들지 않을까 싶은 생각이 드네요.

**Reply by haeseong8880** [reply]
Maria DB, nodejs로 api, 특정시간 크롤링 크론탭 정도 돌리고 있습니다!

매달 5500원 정도만 나와요!

유저는 한달에 3천명정도 밖에 안되는게 아쉽지만요 😅

**Reply by BenjaminKim** [reply]
와.. 디비도 올리고 쓰시는군요. 아주 잘 쓰고 계신 것 같습니다.

**Comment by orangec-at** [comment]
안녕하세요! 어느새인가 제 피드에 나타나셔서 여기까지 왔습니다 :) 제게 좋은 영감이 되고 있어요

궁금한게 있어서 댓글 남기게 되었습니다. 

혹시 실 개발에 사용하고 있는 컴퓨터를(제 상황에서는 m2 pro mac mini) 서버로 사용하면서 개발해도 큰 문제는 없을까요? aws 로 실 운영을 할까 하다가 더 싼 대안은 없을까해서 한번 여쭤봅니다 🤔

**Reply by BenjaminKim** [reply]
m2 맥 미니면 서버로 쓰기에 아주 좋죠.
서버로 쓰면서 개발까지 하는 건 어떤 서비스냐에 따라 다를겁니다.
개발 중이거나 덜 알려진 서비스라면 문제 없을 것 같고요.
돈을 조금이라도 벌고 있는 수준의 서비스라면 그냥 새로 하나 사서 쓰세요. ㅎ

**Reply by orangec-at** [reply]
이렇게 빠르게 답변을 주실 줄이야…!

제 상황은 이제 오픈 전 인프라 구성 단계에 있던 터라, 우선은 홈서버에 구축하는게 좋겠네요 ㅎㅎ

답변해주셔서 감사합니다 🥹

**Comment by KKang29** [comment]
커피한잔 개발자 재호님!! 커피한잔의 시작은 작은 서버였지만 사이즈가 커지면서 사람들이 많이 찾는 서비스가 되가는 모습을 보실 때 기분이 좋으셨겠습니다. 저도... 집에 남는 컴퓨터 및 노트북, 나스로 웹 호스팅 및 서비스를 올려두고 운영하는데 큰 무리는 없더라구요. Ai 서비스라 gpu까지 건드려서 할거는 많은데 오히려 기획, 개발, 디자인, 보안, 최적화 등 a to z로 직접 관리해야히니 힘은 들지만 그 만큼 다양한 정보들도 공부하게 되니까 좋다고 생각합니다! 항상 좋은 글 잘 보고 있습니다. 25년도 좋은 일만 있으시길 바랍니다.

**Reply by BenjaminKim** [reply]
AI서비스까지 돌리시다니... 저보다 더 잘 사용하고 계신걸요? 덕담 감사드리고 @KKang29 님께서도 좋은 일만 있으시길 바라겠습니다.

**Comment by netcitymagazine-boop** [comment]
직접 운영하고 있습니다. 일반 가정용 케이블입니다. 우분투로 운영중에 있고요 도메인과 직접 연결해서 사용하고 있습니다. 가능해요.
